src/fs.rs

```diff
@@ -0,0 +1,441 @@
+
+        if path_remoteness(&lock_file_path) == DirRemoteness::remote {
+            return Err(std::io::Error::new(
+                std::io::ErrorKind::Unsupported,
+                "Directory considered remote. Locking is disabled on remote file systems.",
+            ));
+        }
+
+        // We start by locking the dedicated lock file.
+        // This is required to avoid racing modifications by other threads/processes.
+        // If the lock is not released explicitly, it is released when the file descriptor of
+        // `lock_file` is closed, which happens in the `Drop` implementation of `File`.
+        let lock_file = wopen_cloexec(
+            &lock_file_path,
+            OFlag::O_RDONLY | OFlag::O_CREAT,
+            Mode::S_IRUSR,
+        )?;
+
+        // Try locking the file. Retry if locking was interrupted.
+        while unsafe { flock(lock_file.as_raw_fd(), locking_mode.flock_op()) } == -1 {
+            let err = std::io::Error::last_os_error();
+            if err.kind() != std::io::ErrorKind::Interrupted {
+                return Err(err);
+            }
+        }
+
+        // Open the data file
+        let data_file = wopen_cloexec(file_path, locking_mode.file_flags(), LOCKED_FILE_MODE)?;
+
+        Ok(Self {
+            data_file,
+            _lock_file: lock_file,
+        })
+    }
+
+    pub fn get(&self) -> &File {
+        &self.data_file
+    }
+
+    pub fn get_mut(&mut self) -> &mut File {
+        &mut self.data_file
+    }
+}
+
+/// Runs the `load` function, which should see a consistent state of the file at `path`.
+/// To ensure a consistent state, we use locks to prevent modifications by others.
+/// If locking is unavailable, the `load` function might be executed multiple times,
+/// until it manages a run without the file being modified in the meantime, or until the maximum
+/// number of allowed attempts is reached.
+/// If the file does not exist this function will return an error.
+pub fn lock_and_load<F, UserData>(path: &wstr, load: F) -> std::io::Result<(FileId, UserData)>
+where
+    F: Fn(&File) -> std::io::Result<UserData>,
+{
+    match LockedFile::new(LockingMode::Shared, path) {
+        Ok(locked_file) => {
+            return Ok((
+                file_id_for_file(locked_file.get()),
+                load(locked_file.get())?,
+            ));
+        }
+        Err(e) => {
+            FLOGF!(
+                synced_file_access,
+                "Error trying to lock file '%s': %s",
+                path,
+                e,
+            );
+            // This function might be called when the file does not exist.
+            // Do not abandon locking is such cases.
+            if e.kind() == std::io::ErrorKind::NotFound {
+                // There is no point in continuing in this function if the file does not
+                // exist.
+                return Err(e);
+            } else {
+                FLOG!(synced_file_access, "Disabling locking.");
```

Comment:
this log can go away now as of the last commit.
(That commit should technically come first to make things strictly better in each commit but it's fine..)